File: backend/internal/engine/builtin_rules.go
package engine

import (
	"github.com/suprelory/redact-gateway/pkg/types"
)

// GetBuiltinRules 返回所有内置规则（对齐 TypeScript 版本）
func GetBuiltinRules() []*types.Rule {
	rules := []*types.Rule{
		// ============= 私钥类 (Priority 100) =============
		{
			ID:       "pem-private-key",
			Type:     "PRIVATE_KEY",
			Name:     "PEM 私钥",
			Enabled:  true,
			Builtin:  true,
			Priority: 100,
			Method:   types.MethodRegex,
			Pattern:  `-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----[\s\S]+?-----END[^-]*PRIVATE KEY-----`,
		},

		// ============= API Keys (Priority 90) =============
		{
			ID:       "openai-key",
			Type:     "API_KEY",
			Name:     "OpenAI API Key",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\bsk-[A-Za-z0-9]{20,}`,
		},
		{
			ID:       "anthropic-key",
			Type:     "API_KEY",
			Name:     "Anthropic API Key",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\bsk-ant-api03-[A-Za-z0-9_-]{93}AA`,
		},
		{
			ID:       "github-token",
			Type:     "API_KEY",
			Name:     "GitHub Token",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,255}\b`,
		},
		{
			ID:       "github-fine-grained",
			Type:     "API_KEY",
			Name:     "GitHub Fine-grained PAT",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\bgithub_pat_[A-Za-z0-9_]{50,}`,
		},
		{
			ID:       "gcp-api-key",
			Type:     "API_KEY",
			Name:     "Google Cloud API Key",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\bAIza[0-9A-Za-z_-]{35,38}`,
			AllowList: []string{"AIzaSyabcdefghijklmnopqrstuvwxyz1234567"}, // 示例密钥
		},
		{
			ID:       "aws-access-key",
			Type:     "ACCESS_KEY",
			Name:     "AWS Access Key ID",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\b(?:AKIA|ASIA|ABIA|ACCA)[A-Z0-9]{16}\b`,
			AllowList: []string{"AKIAIOSFODNN7EXAMPLE"}, // AWS 示例密钥
		},
		{
			ID:           "aws-secret-key",
			Type:         "ACCESS_KEY",
			Name:         "AWS Secret Access Key",
			Enabled:      true,
			Builtin:      true,
			Priority:     90,
			Method:       types.MethodRegex,
			Pattern:      `(?i)aws[_-]?secret[_-]?access[_-]?key["']?\s*[:=]\s*["']?([A-Za-z0-9/+=]{40})`,
			CaptureGroup: 1,
		},
		{
			ID:       "alibaba-access-key",
			Type:     "ACCESS_KEY",
			Name:     "阿里云 Access Key",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\bLTAI[A-Za-z0-9]{12,20}`,
		},
		{
			ID:       "jwt-token",
			Type:     "JWT",
			Name:     "JWT Token",
			Enabled:  true,
			Builtin:  true,
			Priority: 90,
			Method:   types.MethodRegex,
			Pattern:  `\beyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}`,
		},

		// ============= 连接字符串 (Priority 85) =============
		{
			ID:       "connection-string",
			Type:     "CONNSTR",
			Name:     "数据库连接字符串",
			Enabled:  true,
			Builtin:  true,
			Priority: 85,
			Method:   types.MethodRegex,
			Pattern:  `\b(?:mysql|postgresql|postgres|mongodb|redis|mssql)://[^\s<>"']+`,
		},

		// ============= 中国 PII (Priority 75-80) =============
		{
			ID:       "china-phone",
			Type:     "PHONE",
			Name:     "中国手机号",
			Enabled:  true,
			Builtin:  true,
			Priority: 80,
			Method:   types.MethodRegex,
			Pattern:  `\b1[3-9]\d{9}\b`,
		},
		{
			ID:       "china-id-card",
			Type:     "IDENTITY",
			Name:     "中国身份证号",
			Enabled:  true,
			Builtin:  true,
			Priority: 75,
			Method:   types.MethodRegex,
			Pattern:  `\b\d{17}[\dXx]\b`,
		},
		{
			ID:       "bank-card",
			Type:     "BANK_CARD",
			Name:     "银行卡号",
			Enabled:  true,
			Builtin:  true,
			Priority: 75,
			Method:   types.MethodRegex,
			Pattern:  `\b\d{13,19}\b`,
		},

		// ============= 邮箱 (Priority 70, 默认禁用) =============
		{
			ID:       "email",
			Type:     "EMAIL",
			Name:     "邮箱地址",
			Enabled:  false, // 默认禁用，误报率高
			Builtin:  true,
			Priority: 70,
			Method:   types.MethodRegex,
			Pattern:  `\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b`,
		},

		// ============= IP 地址 (Priority 65) =============
		{
			ID:        "ipv4-private",
			Type:      "IPPRIVATE",
			Name:      "内网 IPv4 地址",
			Enabled:   true,
			Builtin:   true,
			Priority:  65,
			Method:    types.MethodRegex,
			Pattern:   `\b(?:10|172\.(?:1[6-9]|2\d|3[01])|192\.168)\.\d{1,3}\.\d{1,3}\b`,
			AllowList: []string{"127.0.0.1", "0.0.0.0", "255.255.255.255"},
		},

		// ============= 高熵检测 (Priority 50) =============
		{
			ID:               "high-entropy",
			Type:             "SECRET",
			Name:             "高熵字符串",
			Enabled:          true,
			Builtin:          true,
			Priority:         50,
			Method:           types.MethodEntropy,
			MinLength:        9,
			EntropyThreshold: 5.2,
		},
	}

	return rules
}

// GetRulesByType 按类型过滤规则
func GetRulesByType(ruleType string) []*types.Rule {
	allRules := GetBuiltinRules()
	var filtered []*types.Rule

	for _, rule := range allRules {
		if rule.Type == ruleType {
			filtered = append(filtered, rule)
		}
	}

	return filtered
}

// GetEnabledRules 获取启用的规则
func GetEnabledRules() []*types.Rule {
	allRules := GetBuiltinRules()
	var enabled []*types.Rule

	for _, rule := range allRules {
		if rule.Enabled {
			enabled = append(enabled, rule)
		}
	}

	return enabled
}

// FindRuleByID 根据 ID 查找规则
func FindRuleByID(id string) *types.Rule {
	allRules := GetBuiltinRules()
	for _, rule := range allRules {
		if rule.ID == id {
			return rule
		}
	}
	return nil
}

File: backend/src/engine/builtin-rules.ts
/**
 * 内置规则集
 */

import type { Rule } from '../types.js';

/**
 * 内置规则优先级分配：
 * 100: PEM 私钥（最高危）
 * 90: API Keys / Tokens
 * 85: 连接字符串
 * 80: 手机号
 * 75: 身份证、银行卡
 * 70: 邮箱
 * 65: IP 地址
 * 50: 高熵检测
 */

export const BUILTIN_RULES: Rule[] = [
  // ============= 凭据类 (Priority 90-100) =============

  {
    id: 'pem-private-key',
    type: 'PRIVATE_KEY',
    name: 'PEM 私钥',
    description: '匹配 PEM 格式的私钥（RSA/EC/DSA/OPENSSH）',
    enabled: true,
    builtin: true,
    priority: 100,
    method: 'regex',
    pattern: '-----BEGIN (?:RSA |EC |DSA |OPENSSH |PGP )?PRIVATE KEY-----[\\s\\S]+?-----END[^-]*PRIVATE KEY-----',
    captureGroup: 0,
  },

  {
    id: 'openai-key',
    type: 'API_KEY',
    name: 'OpenAI API Key',
    description: '匹配 sk- 开头的 OpenAI API 密钥',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\bsk-[A-Za-z0-9]{20,}',
    captureGroup: 0,
  },

  {
    id: 'anthropic-key',
    type: 'API_KEY',
    name: 'Anthropic API Key',
    description: '匹配 Anthropic Claude API 密钥',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\bsk-ant-api03-[A-Za-z0-9_-]{93}AA',
    captureGroup: 0,
  },

  {
    id: 'github-token',
    type: 'API_KEY',
    name: 'GitHub Token',
    description: '匹配 GitHub Personal Access Token',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,255}\\b',
    captureGroup: 0,
  },

  {
    id: 'github-fine-grained',
    type: 'API_KEY',
    name: 'GitHub Fine-grained PAT',
    description: '匹配 GitHub Fine-grained Personal Access Token',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\bgithub_pat_[A-Za-z0-9_]{50,}',
    captureGroup: 0,
  },

  {
    id: 'gcp-api-key',
    type: 'API_KEY',
    name: 'Google Cloud API Key',
    description: '匹配 Google Cloud Platform API 密钥',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\bAIza[0-9A-Za-z_-]{35,38}',
    captureGroup: 0,
    allowList: ['AIzaSyabcdefghijklmnopqrstuvwxyz1234567'], // 示例密钥
  },

  {
    id: 'aws-access-key',
    type: 'ACCESS_KEY',
    name: 'AWS Access Key ID',
    description: '匹配 AWS Access Key ID',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\b(?:AKIA|ASIA|ABIA|ACCA)[A-Z0-9]{16}\\b',
    captureGroup: 0,
    allowList: ['AKIAIOSFODNN7EXAMPLE'], // AWS 示例密钥
  },

  {
    id: 'aws-secret-key',
    type: 'ACCESS_KEY',
    name: 'AWS Secret Access Key',
    description: '匹配 AWS Secret Access Key（键值对形式）',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '(?i)aws[_-]?secret[_-]?access[_-]?key["\']?\\s*[:=]\\s*["\']?([A-Za-z0-9/+=]{40})',
    captureGroup: 1,
  },

  {
    id: 'alibaba-access-key',
    type: 'ACCESS_KEY',
    name: '阿里云 Access Key',
    description: '匹配阿里云 Access Key ID',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\bLTAI[A-Za-z0-9]{12,20}',
    captureGroup: 0,
  },

  {
    id: 'jwt-token',
    type: 'JWT',
    name: 'JWT Token',
    description: '匹配 JSON Web Token',
    enabled: true,
    builtin: true,
    priority: 90,
    method: 'regex',
    pattern: '\\beyJ[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}\\.[A-Za-z0-9_-]{8,}',
    captureGroup: 0,
  },

  // ============= 连接字符串 (Priority 85) =============

  {
    id: 'connection-string',
    type: 'CONNSTR',
    name: '数据库连接字符串',
    description: '匹配 MySQL/PostgreSQL/MongoDB/Redis 连接字符串',
    enabled: true,
    builtin: true,
    priority: 85,
    method: 'regex',
    pattern: '\\b(?:mysql|postgresql|postgres|mongodb|redis|mssql):\\/\\/[^\\s<>"\']+',
    captureGroup: 0,
  },

  // ============= 中国 PII (Priority 75-80) =============

  {
    id: 'china-phone',
    type: 'PHONE',
    name: '中国手机号',
    description: '匹配中国大陆手机号码（1 开头 11 位）',
    enabled: true,
    builtin: true,
    priority: 80,
    method: 'regex',
    pattern: '\\b1[3-9]\\d{9}\\b',
    captureGroup: 0,
  },

  {
    id: 'china-id-card',
    type: 'IDENTITY',
    name: '中国身份证号',
    description: '匹配中国大陆身份证号码（18 位带校验）',
    enabled: true,
    builtin: true,
    priority: 75,
    method: 'regex',
    pattern: '\\b\\d{17}[\\dXx]\\b',
    captureGroup: 0,
    // TODO: 添加身份证校验位验证
  },

  {
    id: 'bank-card',
    type: 'BANK_CARD',
    name: '银行卡号',
    description: '匹配 13-19 位银行卡号（含 Luhn 校验）',
    enabled: true,
    builtin: true,
    priority: 75,
    method: 'regex',
    pattern: '\\b\\d{13,19}\\b',
    captureGroup: 0,
    // TODO: 添加 Luhn 校验
  },

  // ============= 邮箱 (Priority 70, 默认禁用) =============

  {
    id: 'email',
    type: 'EMAIL',
    name: '邮箱地址',
    description: '匹配电子邮件地址（默认禁用，误报率高）',
    enabled: false,
    builtin: true,
    priority: 70,
    method: 'regex',
    pattern: '\\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Z|a-z]{2,}\\b',
    captureGroup: 0,
  },

  // ============= IP 地址 (Priority 65) =============

  {
    id: 'ipv4-private',
    type: 'IPPRIVATE',
    name: '内网 IPv4 地址',
    description: '匹配私有 IPv4 地址（10.x, 172.16-31.x, 192.168.x）',
    enabled: true,
    builtin: true,
    priority: 65,
    method: 'regex',
    pattern: '\\b(?:10|172\\.(?:1[6-9]|2\\d|3[01])|192\\.168)\\.\\d{1,3}\\.\\d{1,3}\\b',
    captureGroup: 0,
    allowList: ['127.0.0.1', '0.0.0.0', '255.255.255.255'],
  },

  // ============= 高熵检测 (Priority 50) =============

  {
    id: 'high-entropy',
    type: 'SECRET',
    name: '高熵字符串',
    description: '基于交叉熵的高熵字符串检测（随机密钥/Token）',
    enabled: true,
    builtin: true,
    priority: 50,
    method: 'entropy',
    minLength: 9,
    entropyThreshold: 5.2, // length-aware threshold
  },
];

/**
 * 获取所有启用的规则（按优先级降序）
 */
export function getEnabledRules(rules: Rule[]): Rule[] {
  return rules
    .filter(r => r.enabled)
    .sort((a, b) => b.priority - a.priority);
}

/**
 * 根据 ID 查找规则
 */
export function findRuleById(rules: Rule[], id: string): Rule | undefined {
  return rules.find(r => r.id === id);
}

/**
 * 根据类型查找规则
 */
export function findRulesByType(rules: Rule[], type: string): Rule[] {
  return rules.filter(r => r.type === type);
}
